File: 12-01-2025_Trapping Rain Water.py
class Solution:
    def maxWater(self, arr):
        # code here
        n = len(arr)
        left = [0] * n
        right = [0] * n

        res = 0
        # Fill the left array
        left[0] = arr[0]
        for i in range(1, n):
            left[i] = max(left[i - 1], arr[i])

        # Fill the right array
        right[n - 1] = arr[n - 1]
        for i in range(n - 2, -1, -1):
            right[i] = max(right[i + 1], arr[i])

        # Calculate the accumulated water
        for i in range(1, n - 1):
            min_of_2 = min(left[i - 1], right[i + 1])
            if min_of_2 > arr[i]:
                res += min_of_2 - arr[i]

        return res

File: 12-01-2025_Trapping Rain Water.cpp
class Solution {
  public:
    int maxWater(vector<int> &arr) {
        // code here
         int n = arr.size();
    vector<int> left(n);
    vector<int> right(n);

    int res = 0;
    left[0] = arr[0];
    for (int i = 1; i < n; i++)
        left[i] = max(left[i - 1], arr[i]);
    right[n - 1] = arr[n - 1];
    for (int i = n - 2; i >= 0; i--)
        right[i] = max(right[i + 1], arr[i]);
    for (int i = 1; i < n - 1; i++) {
        int minOf2 = min(left[i - 1], right[i + 1]);
        if (minOf2 > arr[i]) {
            res += minOf2 - arr[i];
        }
    }

    return res;
    }
};